### PR TITLE
Alerting: Fix notification policies inheritance algorithm

### PR DESCRIPTION
**Pull request by gillesdemey**
**What is this feature?**

This PR fixes the inheritance algorithm to compute what properties are inherited from a parent policy in a child policy.

**Additionally** 

This PR normalizes the behaviour of the notification policies API response between Mimir and the built-in Alertmanager / vanilla Alertmanager. 

It also changes to way we write the config for `inherit from parent`, `group_by: []` is only preserved in the configuration by Mimir – other implementations simply return `undefined`.

---

## Note on grouping

There are several ways to define the `group_by` property of a notification policy, the following describes the values and their behaviour.

* `group_by: ['label']` is just regular grouping
* `group_by: []` means "inherit from parent" for child policies
* `group_by: []` means "single group" for the default (root) policy
* `group_by: <undefined>` (the absence of the `group_by` key in the policy) inherits the grouping from the parent route
* `group_by: ['...']` means disable grouping

**Special notes for your reviewer:**

Please review the changes in the inheritance algorithm more closely than the rest of the PR – which is mostly updating the UI to take "no grouping" and "single group" semantics in to account.

I also moved that code [to this file](https://github.com/grafana/grafana/pull/69304/files#diff-226473adaf086beab36f8ac2cd83047a80b10e95657ed68d777debd686df3cf0
)

### CODE REVIEW

File: public/app/features/alerting/unified/components/notification-policies/Policy.tsx [hunk 1]
```diff
@@ -107,10 +107,15 @@ const Policy: FC<PolicyComponentProps> = ({
...
-  const isGrouping = Array.isArray(groupBy) && groupBy.length > 0;
   const hasInheritedProperties = inheritedProperties && Object.keys(inheritedProperties).length > 0;
 
+  const childPolicies = currentRoute.routes ?? [];
+  const inheritedGrouping = hasInheritedProperties && inheritedProperties.group_by;
+  const noGrouping = isArray(groupBy) && groupBy[0] === '...';
+
+  const singleGroup = !noGrouping && isArray(groupBy) && groupBy.length === 0;
+  const customGrouping = !noGrouping && isArray(groupBy) && groupBy.length > 0;
+
```
Comment (soniaAguilarPeiron) [inline public/app/features/alerting/unified/components/notification-policies/Policy.tsx:118]: ☝️ nice!

**Review by soniaAguilarPeiron — APPROVED** [review]
LGTM!🚀

File: public/app/features/alerting/unified/utils/amroutes.ts [hunk 2]
```diff
@@ -183,7 +183,7 @@ export const formAmRouteToAmRoute = (
     receiver,
   } = formAmRoute;
 
-  const group_by = overrideGrouping && groupBy ? groupBy : [];
+  const group_by = overrideGrouping ? (groupBy ? groupBy : []) : undefined;
```
Comment (soniaAguilarPeiron) [inline public/app/features/alerting/unified/utils/amroutes.ts:186]: i would add a comment in here, to make more clear we are setting undefined to inherit from parent.
```suggestion
  const inheritFromParent= undefined;
  const group_by = overrideGrouping ? (groupBy ? groupBy : []) : inheritFromParent;
```

File: public/app/features/alerting/unified/components/notification-policies/Policy.tsx [hunk 3]
```diff
@@ -218,17 +223,27 @@ const Policy: FC<PolicyComponentProps> = ({
...
+                      <Strong>{groupBy.join(', ')}</Strong>
+                    </MetaText>
+                  )}
+                  {singleGroup && (
+                    <MetaText icon="layer-group" data-testid="grouping">
+                      <span>Grouped by</span>
+                      <Strong>Single group</Strong>
+                    </MetaText>
+                  )}
+                  {/* we only want to show "no grouping" on the root policy, children with empty groupBy will inherit from the parent policy */}
```
[the remark this answers is not in the document]
Comment (gillesdemey) [reply public/app/features/alerting/unified/components/notification-policies/Policy.tsx]: Yep, it's no longer accurate – removed

File: public/app/features/alerting/unified/components/Strong.tsx [hunk 4]
```diff
@@ -6,7 +6,7 @@ interface Props {}
 
 const Strong = ({ children }: React.PropsWithChildren<Props>) => {
   const theme = useTheme2();
-  return <strong style={{ color: theme.colors.text.maxContrast }}>{children}</strong>;
+  return <strong style={{ color: theme.colors.text.primary }}>{children}</strong>;
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/components/Strong.tsx:9]: unrelated to the fix, but looks better :)

File: public/app/features/alerting/unified/utils/amroutes.ts [hunk 5]
```diff
@@ -108,8 +108,8 @@ export const amRouteToFormAmRoute = (route: RouteWithID | Route | undefined): Fo
     ],
     continue: route.continue ?? false,
     receiver: route.receiver ?? '',
-    overrideGrouping: Array.isArray(route.group_by) && route.group_by.length !== 0,
-    groupBy: route.group_by ?? [],
+    overrideGrouping: Array.isArray(route.group_by) && route.group_by.length > 0,
+    groupBy: route.group_by ?? undefined,
```
Comment (soniaAguilarPeiron) [inline public/app/features/alerting/unified/utils/amroutes.ts:112]: Could we use the new constant `INHERIT_FROM_PARENT` in here instead of undefined?

**Review by soniaAguilarPeiron — APPROVED** [review]
LGTM!🚀

**Comment by gillesdemey** [comment]
Back-port PR here 👉 https://github.com/grafana/grafana/pull/69782

**Pull request merged**